pkg/agent/controller/traceflow/packetin.go

```diff
@@ -213,11 +213,17 @@ func (c *Controller) parsePacketIn(pktIn *ofctrl.PacketIn) (*opsv1alpha1.Tracefl
 				return nil, nil, err
 			}
 		}
-		if (c.networkConfig.TrafficEncapMode.SupportsEncap() && outputPort == config.DefaultTunOFPort) || outputPort == config.HostGatewayOFPort {
-			// Output port is Tunnel/Gateway port, packet is forwarded.
-			// tunnelDstIP is valid IP in encapMode, and empty string in other modes.
+		gatewayIP := c.nodeConfig.GatewayConfig.IPv4
+		if pktIn.Data.Ethertype == protocol.IPv6_MSG {
+			gatewayIP = c.nodeConfig.GatewayConfig.IPv6
+		}
+		if c.networkConfig.TrafficEncapMode.SupportsEncap() && outputPort == config.DefaultTunOFPort {
 			ob.TunnelDstIP = tunnelDstIP
 			ob.Action = opsv1alpha1.Forwarded
+		} else if ipDst == gatewayIP.String() && outputPort == config.HostGatewayOFPort {
+			ob.Action = opsv1alpha1.Delivered
+		} else if c.networkConfig.TrafficEncapMode.SupportsEncap() && outputPort == config.HostGatewayOFPort {
```

Comment:
Can we check if ipDst in Pod CIDR here? Current we support to trace inter-Node packet without encapsulation, but this behavior is changed in this PR.
Otherwise the e2e test will fail: https://github.com/vmware-tanzu/antrea/pull/1884/checks?check_run_id=1924336861

Reply:
Thanks, I can see that I forgot an if case here. Added it back.